### PR TITLE
Port `margin_ranking_loss` to `TensorIterator`

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72321
* #72324

### CODE REVIEW

**Comment by nkaretnikov** [comment]
going to merge this with other related changes

**Comment by nkaretnikov** [comment]
benchmark: https://github.com/pytorch/pytorch/pull/72327#issuecomment-1029774569

File: aten/src/ATen/native/Loss.cpp [hunk 1]
```diff
@@ -70,6 +74,76 @@ TORCH_IMPL_FUNC(smooth_l1_loss_out)
...
+    ": input1: ", input1.sizes(),
+    ", input2: ", input2.sizes(),
+    ", target: ", target.sizes());
+
+  const auto common_type1 = promoteTypes(input1.scalar_type(), target.scalar_type());
+  const auto common_type2 = promoteTypes(common_type1, input2.scalar_type());
+  const auto reduce = reduction != Reduction::None;
+  const auto options = input1.options().dtype(common_type2);
+
+  // Make sure the result is defined before we use it.
```
Comment (Lezcano) [inline aten/src/ATen/native/Loss.cpp:106]: This comment is not very descriptive. We're doing something a bit odd to avoid having to compute `common_type2` in the out-of-place and the `out=` function. This comment should reflect that.
Usually the pattern would be to create a size zero tensor in the `out-of-place` function and then passing it to the `out` function for it to resize it.

Perhpas a better way to go about it would be to have a look at how other ternary operations handle this (see `addcdiv`), although it may be that they don't do it correctly either. Idk.

File: aten/src/ATen/native/Loss.cpp [hunk 2]
```diff
@@ -70,6 +74,76 @@ TORCH_IMPL_FUNC(smooth_l1_loss_out)
...
+  if (!result.defined()) {
+    result = at::empty(reduce ? IntArrayRef{} : IntArrayRef{0}, options);
+  }
+
+  auto output_iter =
+    reduce ? MaybeOwned<Tensor>::owned(at::empty({0}, options))
```
Comment (Lezcano) [inline aten/src/ATen/native/Loss.cpp:112]: Leave a comment explaining why we need a new tensor in the `reduce`  case, and why otherwise we can use the `result` tensor itself.

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -15345,6 +15401,23 @@ def ref_pairwise_distance(input1, input2):
...
+        sample_inputs_func=sample_inputs_margin_ranking_loss,
+        skips=(
+            # False is not true : Found a sampled tensor of floating-point dtype
+            # torch.float32 sampled with requires_grad=False. If this is
+            # intended, please skip/xfail this test. Remember that sampling
+            # operations are executed under a torch.no_grad contextmanager.
+            DecorateInfo(
+                unittest.expectedFailure, "TestCommon",
+                "test_floating_inputs_are_differentiable",
+                dtypes=(torch.float32,)),
```
Comment (Lezcano) [inline torch/testing/_internal/common_methods_invocations.py:15418]: This needs fixing

**Review by lezcano — COMMENTED** [review]
Very small things needed, but otherwise the PR looks on point!

Could oyu have a look at this one @jbschlosser?

File: tools/autograd/derivatives.yaml [hunk 4]
```diff
@@ -1775,6 +1775,19 @@
   target: smooth_l1_loss_backward(grad, target, self, reduction, beta)
   result: apply_loss_reduction(smooth_l1_loss_backward(self_t.conj(), self_p, target_p, at::Reduction::None, beta).conj() + smooth_l1_loss_backward(target_t.conj(), target_p, self_p, at::Reduction::None, beta).conj(), reduction)
 
+- name: margin_ranking_loss(Tensor input1, Tensor input2, Tensor target, float margin=0.0, int reduction=Mean) -> Tensor
+  input1, input2: "grad.defined() ? sub_backward(margin_ranking_loss_backward_input1(grad, input1, input2, target, margin, reduction)) : std::tuple<Tensor, Tensor>()"
+  target: "grad.defined() ? margin_ranking_loss_backward_target(grad, input1, input2, target, margin, reduction) : Tensor()"
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:1780]: No need to check for grad defined here.

File: tools/autograd/derivatives.yaml [hunk 5]
```diff
@@ -1775,6 +1775,19 @@
   target: smooth_l1_loss_backward(grad, target, self, reduction, beta)
   result: apply_loss_reduction(smooth_l1_loss_backward(self_t.conj(), self_p, target_p, at::Reduction::None, beta).conj() + smooth_l1_loss_backward(target_t.conj(), target_p, self_p, at::Reduction::None, beta).conj(), reduction)
 
+- name: margin_ranking_loss(Tensor input1, Tensor input2, Tensor target, float margin=0.0, int reduction=Mean) -> Tensor
+  input1, input2: "grad.defined() ? sub_backward(margin_ranking_loss_backward_input1(grad, input1, input2, target, margin, reduction)) : std::tuple<Tensor, Tensor>()"
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:1779]: Do you really need input1 and input2 here?
I feel like making this op non-composite is making a lot of things more complex than they need to be?

Comment (Lezcano) [reply tools/autograd/derivatives.yaml:1779]: You need it to recompute the threshold. This could potentially be avoided by computing the threshold in the forward and passing it to the backward functions. It's not clear whether we'll try and benchmark that other approach or not.

This was some sort of introductory PR to TI / derivatives / benchmarking and so on for @nkaretnikov. Benchmarks (https://github.com/pytorch/pytorch/pull/72327#issuecomment-1029774569) show that we managed to make it faster, so there was progress in that direction. Now, it is indeed the case that the code here is much more complex than the previous one, but we got a x2 speed-up, so I guess that's fine :)

**Comment by Lezcano** [comment]
Actually, now that I see the benchmarking code in https://github.com/pytorch/pytorch/pull/72327#issuecomment-1029774569. I see that you're also measuring the creation of the tensors. That's wrong, that should not be included in `f`. Once this is removed, the speed-up should actually be higher than initially reported. The function to test should simply be `torch.nn.functional.margin_ranking_loss(input1, input2, target).backward()`

File: tools/autograd/derivatives.yaml (hunk 5 above)
Comment (albanD) [reply tools/autograd/derivatives.yaml:1779]: Much better speed is definitely a good argument! Note that memory usage is going to be a lot larger though (given all the buffers you save for backward).
I'm happy with the change if we say the speed benefit is worth it!

Comment (albanD) [reply tools/autograd/derivatives.yaml:1779]: Before it was a compositeimplicit op. So not sure what it was saving without checking the code in details.
But if we actually did `y * (input1 - input2)` then we would only save 1 or 2 Tensor of the input size (result of - and y). While the new function saves all 3 now.

**Review by jbschlosser — COMMENTED** [review]
Thanks for the update! Added a few comments below :)

File: aten/src/ATen/native/native_functions.yaml [hunk 6]
```diff
@@ -2854,7 +2854,29 @@
...
+  device_check: NoCheck   # TensorIterator
+  python_module: nn
+  dispatch:
+    CPU, CUDA: margin_ranking_loss_out
+
 - func: margin_ranking_loss(Tensor input1, Tensor input2, Tensor target, float margin=0.0, int reduction=Mean) -> Tensor
+  device_check: NoCheck   # TensorIterator
+  python_module: nn
+  dispatch:
+    CPU, CUDA: margin_ranking_loss
```
[the remark this answers is not in the document]
Comment (Lezcano) [reply aten/src/ATen/native/native_functions.yaml:2867]: We discussed this and we chose not to go that direction because of the hacky way elementwise + reductions are implemented now in structured kernels. In particular, one needs to do things like `maybe_get_output().resize_({});` and create two different TIs in the most used path. I think structured kernels need a better design when dealing with TI, so I would not be surprised if this pattern was changed in the future, so better to do it "the normal way" for now.

[the remark this answers is not in the document]
Comment (jbschlosser) [reply aten/src/ATen/native/native_functions.yaml:2867]: Thanks @Lezcano for the explanation! cc @bdhirsh for visibility on structured kernel complications

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 7]
```diff
@@ -739,11 +739,6 @@ inline Tensor margin_ranking_loss(const Tensor& input1,
                                   const Tensor& target,
                                   double margin,
                                   MarginRankingLossFuncOptions::reduction_t reduction) {
-  TORCH_CHECK(
-    input1.dim() == input2.dim() && input1.dim() == target.dim(),
-    "margin_ranking_loss : All input tensors should have same dimension but got sizes: "
-    "input1: ", input1.sizes(), ", input2: ", input2.sizes(),
-    ", target: ", target.sizes());
```
[the remark this answers is not in the document]
Comment (nkaretnikov) [reply torch/csrc/api/include/torch/nn/functional/loss.h:746]: yes

File: torch/nn/functional.py [hunk 8]
```diff
@@ -3298,7 +3298,7 @@ def margin_ranking_loss(
                 "input1: {}, input2: {}, target: {} ".format(input1.size(), input2.size(), target.size())
             )
         )
-    return torch.margin_ranking_loss(input1, input2, target, margin, reduction_enum)
+    return torch._C._nn.margin_ranking_loss(input1, input2, target, margin, reduction_enum)
```
[the remark this answers is not in the document]
Comment (nkaretnikov) [reply torch/nn/functional.py:3301]: indeed, it broke some test that i had to remove. okay, will do

File: aten/src/ATen/native/PointwiseOps.cpp [hunk 9]
```diff
@@ -41,6 +41,7 @@ TORCH_META_FUNC(addcdiv)
 }
 
 } // namespace meta
+
```
[the remark this answers is not in the document]
Comment (nkaretnikov) [reply aten/src/ATen/native/PointwiseOps.cpp:44]: i did it on purpose since there's usually a newline when a new namespace is introduced, unless it's after another namespace. so i'm in favor of keeping this

File: tools/autograd/derivatives.yaml (hunk 4 above)
Comment (nkaretnikov) [reply tools/autograd/derivatives.yaml:1780]: could you explain why? i think this just follows the pattern that the other ops are using, so when is grad.defined necessary?

File: aten/src/ATen/native/Loss.cpp [hunk 10]
```diff
@@ -70,6 +74,76 @@ TORCH_IMPL_FUNC(smooth_l1_loss_out)
...
+  TORCH_CHECK(
+    input1.sizes() == input2.sizes() &&
+    input1.sizes() == target.sizes(),
+    "margin_ranking_loss: all input tensors should have same size but got sizes"
+    ": input1: ", input1.sizes(),
+    ", input2: ", input2.sizes(),
+    ", target: ", target.sizes());
+
+  const auto common_type1 = promoteTypes(input1.scalar_type(), target.scalar_type());
+  const auto common_type2 = promoteTypes(common_type1, input2.scalar_type());
```
[the remark this answers is not in the document]
Comment (nkaretnikov) [reply aten/src/ATen/native/Loss.cpp:102]: yeah, it's hard for me to judge the impact here, so i'm mostly relying on existing tests and the feedback that others provide here, which is why these names are used too -- to avoid making any incorrect assumptions

[the remark this answers is not in the document]
Comment (nkaretnikov) [reply aten/src/ATen/native/Loss.cpp:102]: if you see a problem with this or have a better idea for names, that's welcome

File: tools/autograd/derivatives.yaml (hunk 4 above)
Comment (Lezcano) [reply tools/autograd/derivatives.yaml:1780]: See the notes at the top of `derivatives.yaml`. There's explained when we need to check for `grad.defined()`.

File: aten/src/ATen/native/Loss.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (Lezcano) [reply aten/src/ATen/native/Loss.cpp:102]: I think it does, as `promoteTypes` is associative (because it is, right? I was thinking about this for some time and I managed to convince myself that it is, but I might be wrong).

File: aten/src/ATen/native/Loss.cpp [hunk 11]
```diff
@@ -70,6 +74,76 @@ TORCH_IMPL_FUNC(smooth_l1_loss_out)
...
+  // https://github.com/pytorch/pytorch/pull/64975
+  TORCH_CHECK(
+    input1.sizes() == input2.sizes() &&
+    input1.sizes() == target.sizes(),
+    "margin_ranking_loss: all input tensors should have same size but got sizes"
+    ": input1: ", input1.sizes(),
+    ", input2: ", input2.sizes(),
+    ", target: ", target.sizes());
+
+  const auto common_type1 = promoteTypes(input1.scalar_type(), target.scalar_type());
```
Comment (ngimel) [inline aten/src/ATen/native/Loss.cpp:101]: why are you doing manual type promotion instead of letting TI take care of it?

Comment (Lezcano) [reply aten/src/ATen/native/Loss.cpp:101]: I didn't know that that could be done. Can you simply set `result` to be of the same type of, say `input1`,  and TI will take care of cast it to the correct type?

Comment (ngimel) [reply aten/src/ATen/native/Loss.cpp:101]: Undefined result will be set to the common type of inputs, dtype of defined result will be respected, but computation will still be done in the common dtype of inputs and copied to result.

File: aten/src/ATen/native/native_functions.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (bdhirsh) [reply aten/src/ATen/native/native_functions.yaml:2867]: > I think structured kernels need a better design when dealing with TI, so I would not be surprised if this pattern was changed in the future, so better to do it "the normal way" for now.

I agree that those ports are a bit of a hack. But some I think the main reason we ended up porting the other loss functions that way are:

(1) We incur some extra overhead by constructing TI twice, but most loss functions aren't *that* perf critical - that extra overhead seems manageable

(2) The "correct fix" is a pretty complicated undertaking. It would probably complicate `TensorIterator` quite a bit - we'd need a way to introduce "intermediate results" to TI. If that doesn't happen in the near future, then we're left without meta tensor support for all loss functions.

(3) The hack is pretty localized, so if we do end up doing (2) at some point, it doesn't seem like a ton of regretful work to refactor the handle of loss functions afterwards. And then we get to avoid blocking anyone that relies on meta tensor support in the meantime.

cc @ngimel

[the remark this answers is not in the document]
Comment (Lezcano) [reply aten/src/ATen/native/native_functions.yaml:2867]: Another way of tackling this would if we added a mechanism (we actually have one) that allowed us to pass the TI from the metakernel to the actual kernel. What would be the problem with that?

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (nkaretnikov) [reply torch/testing/_internal/common_methods_invocations.py:15418]: so, i've looked at this again. i don't understand what i need to do here.

the error message is as follows:

>        #     # False is not true : Found a sampled tensor of floating-point dtype
>        #     # torch.float32 sampled with requires_grad=False. If this is
>        #     # intended, please skip/xfail this test. Remember that sampling
>        #     # operations are executed under a torch.no_grad contextmanager.

i don't understand this at all. so, i'm sampling with `requires_grad=False` because i'm passing `requires_grad` to kwargs. okay.

but here's a docstring of no_grad:

```
class no_grad(_DecoratorContextManager):
    r"""Context-manager that disabled gradient calculation.

    Disabling gradient calculation is useful for inference, when you are sure
    that you will not call :meth:`Tensor.backward()`. It will reduce memory
    consumption for computations that would otherwise have `requires_grad=True`.

    In this mode, the result of every computation will have
    `requires_grad=False`, even when the inputs have `requires_grad=True`.

    This context manager is thread local; it will not affect computation
    in other threads.
```

so it seems like it should be fine with `False`. it should have no effect, no?

is the error msg misleading or something?

i see that this is skipped in a bunch of places too, with a comment saying that a derivative is not implemented for some tensor. but we provide derivatives for all input tensors, no? so what's the issue here?

File: tools/autograd/derivatives.yaml (hunk 4 above)
Comment (nkaretnikov) [reply tools/autograd/derivatives.yaml:1780]: okay, i've tried removing this and now grad tests fail. how exactly do i add undefined tensor support for this op? it seems i need to do it myself, is there a good example that i can follow?

File: aten/src/ATen/native/native_functions.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (Lezcano) [reply aten/src/ATen/native/native_functions.yaml:2867]: cc @ysiraichi with whom we discussed these things.
Also, see https://github.com/pytorch/pytorch/issues/70678 which was written after the discussion.

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:15418]: You are returning a tensor with  `requires_grad=False` from your sampling function. You need to return them with `requires_grad=requires_grad`, otherwise the autograd will not be tested from them. If you do operation son them (e.g. `sign`...) they will lose the `requires_grad` attribute you created them with, as sampling functions are executed under a `no_grad` context manager.

File: aten/src/ATen/native/Loss.cpp (hunk 11 above)
Comment (Lezcano) [reply aten/src/ATen/native/Loss.cpp:101]: I am not 100% sure, as I also remember to have problems passing an undefined tensor to TI. Perhaps @ngimel can help here?

Comment (nkaretnikov) [reply aten/src/ATen/native/Loss.cpp:101]: okay i'll take a look, ty

**Comment by ngimel** [comment]
Can we go back to motivation for this PR? 
Pros:
- it improves margin_ranking_loss performance
Neutral:
- Performance of loss is rarely a perf bottleneck, so improving it won't show in e2e benchmarks
Cons:
- Increases code complexity (+412/-36 lines)
- Has some unsolved technical problems (manual type promotion is an anti-pattern)
- Margin loss used to work for meta tensors because it was a composite op consisting of ops that use structured kernels, now it won't. 
- Similarly, it used to be CompositeImplicitAutograd op that other backends could just use, now it is dispatched to CUDA/CPU, so other backends have to reimplement it
- Increases build time/cuda context size (not by much, but still)
If this was a perf-critical operation, the cons would be offset by a very important pro of better perf. But I'm not sure this is the case. 
cc @bdhirsh for the point about structured.

**Comment by Lezcano** [comment]
I do agree with the "Increases code complexity (+412/-36 lines)" point. All the other points in the Cons can be fixed. 

Now, there are quite a few losses implemented via TI. See for example the following PR from @jbschlosser:
https://github.com/pytorch/pytorch/pull/50553
This PR is just one more going in that direction.

**Comment by jbschlosser** [comment]
Yeah, #50553 was an onboarding task for me and definitely isn't a great shining example of best practices (unfortunately) :)

**Comment by nkaretnikov** [comment]
@Lezcano so what do we do with this one? do we fix the cons or do we drop this entirely?

**Comment by Lezcano** [comment]
What do you say @jbschlosser @ngimel ?
For what is worth, I did mention this to Joel before starting to work on this.

**Comment by Lezcano** [comment]
For reference, I think that merging this after fixing the points raised by @ngimel about broadcasting and metakernels would be fine.

Now, if we choose to go down the not merging road, then I think that, for consistency, mse, huber loss and the others should also be implemented as composite functions.

In my opinion, I think the best option is to go this way for now. In the future, we could move from here to having a template to implement loss functions with minimal boilerplate in a uniform way. That way we can have somewhat compact and correct implementations that are also fast.
